shift_proposal.h

```diff
@@ -2,33 +2,128 @@
 #define SHIFT_PROPOSAL_H
 
 #include <algorithm>
-#include <cassert>
 #include <iterator>
-#include <optional>
+#include <type_traits>
 #include <utility>
 
+template<class I>
+using difference_type_t = typename std::iterator_traits<I>::difference_type;
+
+template<class I>
+using iterator_category_t = typename std::iterator_traits<I>::iterator_category;
+
+template<class I, class Tag, class = void>
+constexpr bool is_category = false;
+template<class I, class Tag>
+constexpr bool is_category<I, Tag, std::enable_if_t<
+    std::is_convertible_v<iterator_category_t<I>, Tag>>> = true;
+
+/// Increment (decrement for negative n) i |n| times or until i == bound,
+/// whichever comes first. Returns n + the difference between i's final position
+/// and its initial position. (Note: "advance" has overloads with this behavior
+/// in the Ranges TS.)
```

Comment:
Yes. I rephrased this twice and apparently stopped in the middle of a third time.